### ISSUE TITLE
```grid_sample``` with 'circular' border conditions

### ISSUE THREAD

**Issue by rossimattia** [pytorch/pytorch#24089]
In ```torch.nn.functional.grid_sample```, it would be nice to have the circular border conditions implemented as well.

**Comment by jerryzh168** [comment]
Could you provide more details? Also, feel free to submit a PR

**Comment by SamPruden** [comment]
I'll second this request. Circular padding is commonly used in graphics applications on the GPU anywhere where you have a small tiled texture. That can be as simple as literal tiles on a floor or as abstract as tiled noise textures used to generate clouds.

`grid_sample` comes close to providing the same features as GPU texture sampling with the normalised coordinates and linear interpolation, but the missing circular padding makes it an incomplete tool to mimic the GPU behaviour. It can be accomplished by manually padding and then adjusting every sampling coordinate, but that's fiddly and inefficient.

For those of us working on problems related to graphics shaders, this would be a very nice addition.

**Comment by MasterHow** [comment]
This requirement may also be encountered when processing panoramas.

**Comment by chengzegang** [comment]
in some cases, especially for coordinate system transformation (from Cartesian to polar, vise versal),
you want the "circular" effect when dealing with those edges.

i quickly wrote a sample code for such uses:
```

import math
from typing import Tuple
from torch import Tensor
import torch
import torch.nn.functional as F


@torch.jit.script
def cartesian_to_polar(x: Tensor, y: Tensor, z: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
    r = torch.sqrt(x**2 + y**2 + z**2)
    theta = torch.acos(z / r)
    phi = torch.atan2(y, x)
    return r, theta, phi


@torch.jit.script
def polar_to_cartesian(r: Tensor, theta: Tensor, phi: Tensor) -> Tuple[Tensor, Tensor, Tensor]:
    # r: 0~1
    # theta: -pi/2~pi/2
    # phi: -pi~pi
    x = r * torch.cos(phi) * torch.cos(theta)
    y = r * torch.cos(phi) * torch.sin(theta)
    z = r * torch.sin(phi) * torch.sign(phi)

    return x, y, z


@torch.jit.script
def view_as_polar(
    pointcloud: Tensor,
    out_shape: Tuple[int, int, int] = (256, 256, 256),
    mode: str = "nearest",
    padding_mode: str = "zeros",
    align_corners: bool = True,
):
    r, theta, phi = torch.meshgrid(
        torch.linspace(0, 1, out_shape[0]),
        torch.linspace(-torch.pi, torch.pi, out_shape[1]),
        torch.linspace(-torch.pi / 2, torch.pi / 2, out_shape[2]),
        indexing="ij",
    )
    x, y, z = polar_to_cartesian(r, theta, phi)
    grid = torch.stack([x, y, z], dim=-1)
    grid = grid.unsqueeze(0).expand(pointcloud.shape[0], -1, -1, -1, -1)

    polar_view = F.grid_sample(pointcloud, grid, mode=mode, padding_mode=padding_mode, align_corners=align_corners)
    return polar_view


@torch.jit.script
def view_as_cartesian(
    pointcloud: Tensor,
    out_shape: Tuple[int, int, int] = (256, 256, 256),
    mode: str = "nearest",
    padding_mode: str = "zeros",
    align_corners: bool = False,
    scale: float = 1.0,
):
    x, y, z = torch.meshgrid(
        torch.linspace(-1, 1, out_shape[0], device=pointcloud.device, dtype=torch.float32),
        torch.linspace(-1, 1, out_shape[1], device=pointcloud.device, dtype=torch.float32),
        torch.linspace(-1, 1, out_shape[2], device=pointcloud.device, dtype=torch.float32),
        indexing="ij",
    )
    r, theta, phi = cartesian_to_polar(x, y, z)
    r_max = math.sqrt(3)
    r = r * 2 / r_max - 1
    theta = theta * 2 / torch.pi - 1
    phi = phi / torch.pi
    r = r * scale

    grid = torch.stack([r, theta, phi], dim=-1)
    grid = grid.unsqueeze(0).expand(pointcloud.shape[0], -1, -1, -1, -1)

    polar_view = F.grid_sample(
        pointcloud.type(torch.float32), grid, mode=mode, padding_mode=padding_mode, align_corners=align_corners
    ).type_as(pointcloud)
    return polar_view

```

in function ```view_as_cartesian```, it makes senses to pad data "circularly" or "rotationally", just like walking on the Earth forever won't fly you to the Mars.

***note***: the ```view_as_polar``` samples to a 3d-sphere with radius 1 while the ```view_as_cartesian``` samples to a 3d-cube with edge size 1, so they are not inverse transform to each other.

i didn't check my code correctness but it is sufficient enough to understand my idea.
